lib/srv/desktop/rdp/rdpclient/src/client.rs

```diff
@@ -40,6 +40,33 @@ use crate::cliprdr::{ClipboardFn, TeleportCliprdrBackend};
 use crate::rdpdr::scard::SCARD_DEVICE_ID;
 use crate::rdpdr::TeleportRdpdrBackend;
 
+#[macro_use]
+mod macros {
+    macro_rules! get_svc_processor {
+        ($x224_processor:expr, $svc_processor:ty) => {
+            $x224_processor
+                .get_svc_processor::<$svc_processor>()
+                .ok_or(ClientError::InternalError)?
+        };
+    }
+
+    macro_rules! get_svc_processor_mut {
+        ($x224_processor:expr, $svc_processor:ty) => {
+            $x224_processor
+                .get_svc_processor_mut::<$svc_processor>()
+                .ok_or(ClientError::InternalError)?
+        };
+    }
+
+    macro_rules! get_backend_mut {
+        ($x224_processor:expr, $svc_processor:ty, $backend:ty) => {
+            get_svc_processor_mut!($x224_processor, $svc_processor)
+                .downcast_backend_mut::<$backend>()
+                .ok_or(ClientError::InternalError)?
+        };
+    }
+}
```

Comment:
Can't we turn that into regular function? It doesn't look like we do anything that requires macro here?

Reply:
We can change `get_svc_processor` and `get_svc_processor_mut` to functions: bffa13083100389426d5db27e441d63d0a954eb7

For `get_backend_mut`, I don't think it's possible. Basically, there's no way to parameterize a function such that we can call `downcast_backend_mut` from within it, even if we change IronRDP:

`downcast_backend_mut` is specific function on [`Cliprdr`](https://github.com/Devolutions/IronRDP/blob/4cf5db28ca7e319c23e5d42ec1604cb3397b306f/crates/ironrdp-cliprdr/src/lib.rs#L86-L88) (and `Rdpdr`) which downcasts it to a generic `T: RdpdrBackend`, meaning the function signature returns a generic (`Option<&mut T>`). In IronRDP, we use `StaticVirtualChannelProccesor`/`CliprdrBackend`/`RdpdrBackend` in `Box` types like `Box<&dyn CliprdrBackend>`, which constrains `CliprdrBackend` and co to being [object safe](https://doc.rust-lang.org/reference/items/traits.html#object-safety). One of those constraints is that none of the trait functions can have any type parameters:

> Not have any type parameters (although lifetime parameters are allowed)

This means that there's no way to change the `StaticVirtualChannelProccesor`/`CliprdrBackend`/`RdpdrBackend` traits such that they return an `Option<&mut T>`, which means that there's no way (that I can see) to parameterize a function to give us back what we need (e.g. `TeleportRdpdrBackend` or `TeleportCliprdrBackend`).

Feel free to show me up though, maybe there's something that I missed.

Reply:
Can we just don't use generic functions/macros? We have 2 backends so maybe add 2 non-generic function?

```rust 
fn cliprdr_backend(
    x224_processor: &mut X224Processor,
) -> ClientResult<&mut TeleportCliprdrBackend> {
    x224_processor
        .get_svc_processor_mut::<Cliprdr>()
        .and_then(|c| c.downcast_backend_mut::<TeleportCliprdrBackend>())
        .ok_or(ClientError::InternalError)
}

fn rdpdr_backend(
    x224_processor: &mut X224Processor,
) -> ClientResult<&mut TeleportRdpdrBackend> {
    x224_processor
        .get_svc_processor_mut::<Rdpdr>()
        .and_then(|c| c.downcast_backend_mut::<TeleportRdpdrBackend>())
        .ok_or(ClientError::InternalError)
}
...
let mut x224_processor = Self::x224_lock(&x224_processor)?;
let rdpdr = Self::rdpdr_backend(&mut x224_processor)?;
rdpdr.handle_tdp_sd_info_response(res)?;
```

Reply:
Yeah good idea: 1919c42c290327d84a417d5d3d57ec6b639ad778